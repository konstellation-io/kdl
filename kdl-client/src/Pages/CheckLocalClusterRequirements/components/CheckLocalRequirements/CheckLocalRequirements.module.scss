@import 'Styles/variables';
@import 'Styles/mixins';
@import 'Styles/colors';

$circle-diameter: $grid-unit;
$circle-top: 11px;
$circle-left: 0px;

.container {
  margin-bottom: 6 * $grid-unit;
}

.checkList ul {
  @include font-small;
  color: font-color(lowlight);

  padding: 0;

  li {
    transition: color ease 0.6s;
  }
}

@mixin status-circle($color) {
  color: $color;

  &:after {
    background-color: $color;
  }

  &:before {
    background-color: $color;
  }
}

.circle {
  @include circle(transparent);

  content: '';
  position: absolute;
  left: 0;
  top: 11px;
}

.requirement {
  list-style: none;
  position: relative;
  padding: $grid-unit 0 $grid-unit 2 * $grid-unit;

  &:before {
    @extend .circle;
  }

  &:after {
    @extend .circle;
  }

  &.pending {
    @include status-circle(palette(highlight));

    &:before {
      animation: loading-anim 2s ease 0s infinite;
    }
  }

  &.fail:before,
  &.ok:before {
    animation: loading-anim 0s ease, loading-anim 1s ease;
  }
  &.fail {
    @include status-circle(palette(error));
  }
  &.ok {
    @include status-circle(palette(success));
  }
}

.errorBox {
  position: relative;
  margin: 1.5 * $grid-unit 0 0 4 * $grid-unit;

  &:before,
  &:after {
    content: '';
    position: absolute;
  }

  &:before {
    top: -24px;
    left: -28px;
    width: 14px;
    height: 43px;
    border: dashed 1px palette(error);
    border-right-color: transparent;
    border-top-color: transparent;
  }
  &:after {
    width: 10px;
    height: 10px;
    border: 1px solid palette(error);
    border-top-color: transparent;
    border-left-color: transparent;
    top: 13px;
    left: -21px;
    transform: rotate(-33deg) skew(25deg);
  }
}

@mixin set-anim-state($diameter, $offset, $opacity) {
  width: $diameter;
  height: $diameter;
  top: $circle-top - $offset;
  left: $circle-left - $offset;
  opacity: $opacity;
}

$circle-increase: 1.8 * $grid-unit;
$circle-diameter-end: $circle-diameter + $circle-increase;
@keyframes loading-anim {
  0% {
    @include set-anim-state($circle-diameter, 0px, 0.6);
  }
  60%,
  100% {
    @include set-anim-state($circle-diameter-end, $circle-increase / 2, 0);
  }
}
