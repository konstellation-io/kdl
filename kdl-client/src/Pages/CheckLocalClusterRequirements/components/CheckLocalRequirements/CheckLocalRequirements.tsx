import React, { useEffect, useState } from 'react';

import { CheckState } from 'Pages/CheckLocalClusterRequirements/CheckLocalClusterRequirements';
import { cloneDeep } from 'lodash';
import cx from 'classnames';
import { ipcRenderer } from 'electron';
import styles from './CheckLocalRequirements.module.scss';

enum CheckId {
  k8s = 'k8s',
  minikube = 'minikube',
  helm = 'helm',
}

type CheckRequirementResponse = [CheckId, boolean];

type Check = {
  id: CheckId;
  label: string;
  isOk: boolean | null;
};

const generateInitialCheck = (id: CheckId, label: string) => ({
  id,
  label,
  isOk: null,
});

const initialCheckStatus: Check[] = [
  generateInitialCheck(CheckId.k8s, 'Kubernetes installed'),
  generateInitialCheck(CheckId.minikube, 'Minikube running'),
  generateInitialCheck(CheckId.helm, 'Helm available'),
];

type Props = {
  setChecksState: (newCheckState: CheckState) => void;
};

function CheckLocalRequirements({ setChecksState }: Props) {
  const [checks, setChecks] = useState(initialCheckStatus);

  useEffect(() => {
    const onRequirementReply = (
      _: unknown,
      [requirement, isOk]: CheckRequirementResponse
    ) => {
      setChecks((actChecks) => {
        const newCheckStatus = cloneDeep(actChecks);
        const actCheck = newCheckStatus.find((e) => e.id === requirement);
        if (actCheck) actCheck.isOk = isOk;

        return newCheckStatus;
      });
    };

    ipcRenderer.on('checkRequirementReply', onRequirementReply);

    checks.forEach((check) => {
      ipcRenderer.send('runCheckRequirement', check.id);
    });

    return () => {
      ipcRenderer.removeListener('checkRequirementReply', onRequirementReply);
    };
    // We want to execute this on on component mount/unmount
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    switch (true) {
      case checks.some((c) => c):
        // FIXME: Uncomment next line and remove previous.
        // case checks.every((c) => c.isOk):
        setChecksState(CheckState.OK);
        break;
      case checks.some((c) => c.isOk === false):
        setChecksState(CheckState.ERROR);
        break;
    }
  }, [checks, setChecksState]);

  const requirements = checks.map((check) => (
    <li
      key={check.id}
      className={cx(styles.requirement, {
        [styles.pending]: check.isOk === null,
        [styles.ok]: check.isOk,
        [styles.fail]: check.isOk === false,
      })}
    >
      {check.label}
    </li>
  ));

  return (
    <div className={styles.container}>
      <div className={styles.checkList}>
        <ul>{requirements}</ul>
      </div>
    </div>
  );
}

export default CheckLocalRequirements;
