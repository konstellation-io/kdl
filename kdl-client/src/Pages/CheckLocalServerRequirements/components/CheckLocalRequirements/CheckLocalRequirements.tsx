import React, { Fragment, useEffect, useState } from 'react';

import { CheckState } from 'Pages/CheckLocalServerRequirements/CheckLocalServerRequirements';
import ErrorBox from 'Components/ErrorBox/ErrorBox';
import { cloneDeep } from 'lodash';
import cx from 'classnames';
import { ipcRenderer } from 'electron';
import styles from './CheckLocalRequirements.module.scss';

enum CheckId {
  k8s = 'k8s',
  minikube = 'minikube',
  helm = 'helm',
}

const checkError = {
  k8s: {
    title: 'Check 1 error',
    message: 'Check 1 message',
    docUrl: 'google.es',
  },
  minikube: {
    title: 'Check 2 error',
    message: 'Check 2 message',
    docUrl: 'google.es',
  },
  helm: {
    title: 'Check 3 error',
    message: 'Check 3 message',
    docUrl: 'google.es',
  },
};

type CheckRequirementResponse = [CheckId, boolean];

type Check = {
  id: CheckId;
  label: string;
  isOk: boolean | null;
  open: boolean;
};

const generateInitialCheck = (id: CheckId, label: string) => ({
  id,
  label,
  isOk: null,
  open: false,
});

const initialCheckStatus: Check[] = [
  generateInitialCheck(CheckId.k8s, 'Kubernetes installed'),
  generateInitialCheck(CheckId.minikube, 'Minikube running'),
  generateInitialCheck(CheckId.helm, 'Helm available'),
];

type Props = {
  setChecksState: (newCheckState: CheckState) => void;
};

function CheckLocalRequirements({ setChecksState }: Props) {
  const [checks, setChecks] = useState(initialCheckStatus);

  function resetCheck(checkId: string) {
    const newChecks = cloneDeep(checks);
    const target = newChecks.find((c) => c.id === checkId);
    if (target) target.isOk = null;

    setChecks(newChecks);
  }

  function checkRequirement(checkId: string) {
    resetCheck(checkId);
    ipcRenderer.send('checkRequirement', checkId);
  }

  useEffect(() => {
    const onRequirementReply = (
      _: unknown,
      [requirement, isOk]: CheckRequirementResponse
    ) => {
      setChecks((actChecks) => {
        const newCheckStatus = cloneDeep(actChecks);
        const actCheck = newCheckStatus.find((e) => e.id === requirement);
        if (actCheck) actCheck.isOk = isOk;

        return newCheckStatus;
      });
    };

    ipcRenderer.on('checkRequirementReply', onRequirementReply);

    checks.forEach((check) => {
      checkRequirement(check.id);
    });

    return () => {
      ipcRenderer.removeListener('checkRequirementReply', onRequirementReply);
    };
    // We want to execute this on on component mount/unmount
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    if (checks.every((c) => c.isOk)) {
      setChecksState(CheckState.OK);
    } else if (checks.some((c) => c.isOk === false)) {
      setChecksState(CheckState.ERROR);
    }
  }, [checks, setChecksState]);

  function toggleCheckOpen(idx: number, isOpening: boolean) {
    const newChecks = cloneDeep(checks);

    // Close all checks but opened
    newChecks.forEach((check) => (check.open = false));
    if (isOpening) newChecks[idx].open = true;

    setChecks(newChecks);
  }

  const requirements = checks.map((check, idx) => (
    <Fragment key={check.id}>
      <li
        className={cx(styles.requirement, {
          [styles.pending]: check.isOk === null,
          [styles.ok]: check.isOk,
          [styles.fail]: check.isOk === false,
        })}
      >
        {check.label}
      </li>
      {check.isOk === false && (
        <div className={styles.errorBox}>
          <ErrorBox
            title={checkError[check.id].title}
            message={checkError[check.id].message}
            docUrl={checkError[check.id].docUrl}
            onChange={() => toggleCheckOpen(idx, !check.open)}
            openController={check.open}
            action={{
              label: 'RETRY',
              onClick: () => checkRequirement(check.id),
            }}
          />
        </div>
      )}
    </Fragment>
  ));

  return (
    <div className={styles.container}>
      <div className={styles.checkList}>
        <ul>{requirements}</ul>
      </div>
    </div>
  );
}

export default CheckLocalRequirements;
