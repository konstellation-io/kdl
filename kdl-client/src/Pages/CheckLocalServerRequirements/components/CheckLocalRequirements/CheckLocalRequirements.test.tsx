import CheckLocalRequirements from './CheckLocalRequirements';
import React from 'react';
import { shallow } from 'enzyme';

const setChecksStateMock = jest.fn();

const component = shallow(
  <CheckLocalRequirements setChecksState={setChecksStateMock} />
);

describe('CheckLocalRequirements page', () => {
  test('Component match snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  test('initially all checks are pending', () => {
    expect(component.find('.pending').length).toBe(3);
  });

  // TODO: study how to test ipcRender messages and useEffect tests
});
