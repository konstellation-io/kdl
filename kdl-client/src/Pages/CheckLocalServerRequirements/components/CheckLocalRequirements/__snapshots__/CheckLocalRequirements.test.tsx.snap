// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckLocalRequirements page Component match snapshot 1`] = `
<div
  className="container"
>
  <div
    className="checkList"
  >
    <ul>
      <li
        className="requirement pending"
      >
        Kubernetes installed
      </li>
      <li
        className="requirement pending"
      >
        Minikube running
      </li>
      <li
        className="requirement pending"
      >
        Helm available
      </li>
    </ul>
  </div>
</div>
`;
